mmcv/ops/csrc/common/pytorch_device_registry.hpp

```diff
@@ -0,0 +1,135 @@
+    str.append(std::to_string(device.index()));
+  }
+  return str;
+}
+
+// Registry
+template <typename F, F f>
+class DeviceRegistry;
+
+template <typename Ret, typename... Args, Ret (*f)(Args...)>
+class DeviceRegistry<Ret (*)(Args...), f> {
+ public:
+  using FunctionType = Ret (*)(Args...);
+
+  void Register(at::DeviceType device, FunctionType function) {
+    entries_.insert({device, function});
+  }
+
+  FunctionType Find(at::DeviceType device) const {
+    auto it = entries_.find(device);
+    if (it != entries_.end()) {
+      return it->second;
+    }
+    return nullptr;
+  }
+
+  static DeviceRegistry& instance() {
+    static DeviceRegistry inst;
+    return inst;
+  }
+
+ private:
+  DeviceRegistry() = default;
+  std::map<at::DeviceType, FunctionType> entries_;
+};
+
+// get device of first tensor param
+
+template <typename T, typename... Args,
+          std::enable_if_t<std::is_same<std::decay_t<T>, at::Tensor>::value,
+                           bool> = true>
+at::Device GetFirstTensorDevice(T&& t, Args&&... args) {
+  return std::forward<T>(t).device();
+}
+template <typename T, typename... Args,
+          std::enable_if_t<!std::is_same<std::decay_t<T>, at::Tensor>::value,
+                           bool> = true>
+at::Device GetFirstTensorDevice(T&& t, Args&&... args) {
+  return GetFirstTensorDevice(std::forward<Args>(args)...);
+}
+
+// check device consistency
+
+inline std::pair<int, at::Device> CheckDeviceConsistency(
+    const at::Device& device, int index) {
+  return {index, device};
+}
+
+template <typename T, typename... Args,
+          std::enable_if_t<!std::is_same<std::decay_t<T>, at::Tensor>::value,
+                           bool> = true>
+std::pair<int, at::Device> CheckDeviceConsistency(const at::Device& device,
+                                                  int index, T&& t,
+                                                  Args&&... args);
+
+template <typename T, typename... Args,
+          std::enable_if_t<std::is_same<std::decay_t<T>, at::Tensor>::value,
+                           bool> = true>
+std::pair<int, at::Device> CheckDeviceConsistency(const at::Device& device,
+                                                  int index, T&& t,
+                                                  Args&&... args) {
+  auto new_device = std::forward<T>(t).device();
+  if (new_device.type() != device.type() ||
+      new_device.index() != device.index()) {
+    return {index, new_device};
+  }
+  return CheckDeviceConsistency(device, index + 1, std::forward<Args>(args)...);
+}
+
+template <
+    typename T, typename... Args,
+    std::enable_if_t<!std::is_same<std::decay_t<T>, at::Tensor>::value, bool>>
+std::pair<int, at::Device> CheckDeviceConsistency(const at::Device& device,
+                                                  int index, T&& t,
+                                                  Args&&... args) {
+  return CheckDeviceConsistency(device, index + 1, std::forward<Args>(args)...);
+}
```

Comment:
The implmentation of `CheckDeviceConsistency` is quite involved and may hurt compile time, may be we can first store the devices in a vector and then check for consistency. It shoud be more compile time friendly and is able to report all inconsistent args, like

```C++
// assume this is at::Tensor
struct Tensor { int device; };

inline int GetDevice(const Tensor &t) { return t.device; }
// return an invalid device value for non-Tensor type, or explicitly
// use std::pair to indicate the validity in case the semantic for
// DeviceType changes in Torch
constexpr int GetDevice(...) { return -1; }

// put the definition in a cpp file 
int GetConsistentDeviceImpl(const std::vector<int>& devices) {
  auto expected = std::find_if(begin(devices), end(devices), [](int type) { return type >= 0; });
  // no Tensor in the args, can't determine device type
  assert(expected != end(devices));
  for (size_t i = 0; i < devices.size(); ++i) {
    // flag inconsistency of i-th arg
    assert(devices[i] == *expected || devices[i] < 0);
  }
  return *expected;
}

template<class...Args>
int GetConsistentDevice(Args &&...args) {
  std::vector<int> devices{GetDevice(std::forward<Args>(args))...};
  return GetConsistentDeviceImpl(devices);
}
```